examples/generative/finetune_stable_diffusion.py

```diff
@@ -0,0 +1,507 @@
+
+    def sample_from_encoder_outputs(self, outputs):
+        mean, logvar = tf.split(outputs, 2, axis=-1)
+        logvar = tf.clip_by_value(logvar, -30.0, 20.0)
+        std = tf.exp(0.5 * logvar)
+        sample = tf.random.normal(tf.shape(mean), dtype=mean.dtype)
+        return mean + std * sample
+
+    def save_weights(self, filepath, overwrite=True, save_format=None, options=None):
+        # Overriding this method will allow us to use the `ModelCheckpoint`
+        # callback directly with this trainer class. In this case, it will
+        # only checkpoint the `diffusion_model` since that's what we're training
+        # during fine-tuning.
+        self.diffusion_model.save_weights(
+            filepath=filepath,
+            overwrite=overwrite,
+            save_format=save_format,
+            options=options,
+        )
+
+
+"""
+One important implementation detail to note here. Instead of directly taking
+the latent produced by the image encoder (which is a VAE), we sample from the
+mean and log-variance predicted by it. This way, we can achieve better sample
+quality and diversity. 
+
+It's common to add support for mixed-precision training along with exponential
+moving averaging of model weights for fine-tuning these models. However, in the interest
+of brevity, we discarded those elements. More on this later in the tutorial.
+"""
+
+"""
+## Initialize the trainer and compile it
+"""
+
+# Enable mixed-precision training if the underlying GPU has tensor cores.
+USE_MP = True
+if USE_MP:
+    keras.mixed_precision.set_global_policy("mixed_float16")
+
+image_encoder = ImageEncoder(RESOLUTION, RESOLUTION)
+diffusion_ft_trainer = Trainer(
+    diffusion_model=DiffusionModel(RESOLUTION, RESOLUTION, MAX_PROMPT_LENGTH),
+    # Remove the top layer from the encoder, which cuts off the variance and only
+    # returns the mean.
+    vae=tf.keras.Model(
+        image_encoder.input,
+        image_encoder.layers[-2].output,
+    ),
+    noise_scheduler=NoiseScheduler(),
+    use_mixed_precision=USE_MP,
+)
+
+# These hyperparameters come from this tutorial by Hugging Face:
+# https://huggingface.co/docs/diffusers/training/text2image
+lr = 1e-5
+beta_1, beta_2 = 0.9, 0.999
+weight_decay = (1e-2,)
+epsilon = 1e-08
+
+optimizer = tf.keras.optimizers.experimental.AdamW(
+    learning_rate=lr,
+    weight_decay=weight_decay,
+    beta_1=beta_1,
+    beta_2=beta_2,
+    epsilon=epsilon,
+)
+diffusion_ft_trainer.compile(optimizer=optimizer, loss="mse")
+
+"""
+## Fine-tuning
+
+To keep the runtime of this tutorial short, we just fine-tune for an epoch.
+"""
+
+epochs = 1
+ckpt_path = "finetuned_stable_diffusion.h5"
+ckpt_callback = tf.keras.callbacks.ModelCheckpoint(
+    ckpt_path,
+    save_weights_only=True,
+    monitor="loss",
+    mode="min",
+)
+diffusion_ft_trainer.fit(training_dataset, epochs=epochs, callbacks=[ckpt_callback])
+
+"""
+## Inference
+
+We fine-tuned the model for 60 epochs on an image resolution of 512x512. To allow
```

Comment:
IIUC -- the above 1 epoch training is using 256x256, but users at this point will load pre-fine-tuned 512x512 weights?

Reply:
Yes.